### PR TITLE
Feature/podcast album

### PR DESCRIPTION
**Pull request by Ponchimeow**

### CODE REVIEW

**Comment by Ponchimeow** [comment]
[(5) 【音頻專輯】音頻專輯後台切版+資料串接 [4]](https://trello.com/c/WeoNQigz/2455-5-%E3%80%90%E9%9F%B3%E9%A0%BB%E5%B0%88%E8%BC%AF%E3%80%91%E9%9F%B3%E9%A0%BB%E5%B0%88%E8%BC%AF%E5%BE%8C%E5%8F%B0%E5%88%87%E7%89%88%E8%B3%87%E6%96%99%E4%B8%B2%E6%8E%A5-4)

File: src/components/admin/AdminMenu.tsx [hunk 1]
```diff
@@ -172,6 +171,16 @@ const AdminMenu: React.FC<MenuProps> = ({ children, ...menuProps }) => {
           key: 'podcast_program_category',
           name: formatMessage(commonMessages.menu.podcastCategory),
         },
+        {
+          permissionIsAllowed: true,
```
Comment (futianshen) [inline src/components/admin/AdminMenu.tsx:175]: `app-owner` only

File: src/components/admin/AdminMenu.tsx [hunk 2]
```diff
@@ -172,6 +171,16 @@ const AdminMenu: React.FC<MenuProps> = ({ children, ...menuProps }) => {
...
+          key: 'podcast_album_collection',
+          name: formatMessage(commonMessages.menu.podcastAlbum),
+        },
+        {
+          permissionIsAllowed: true,
```
Comment (futianshen) [inline src/components/admin/AdminMenu.tsx:180]: same as above

File: src/components/admin/AdminMenu.tsx [hunk 3]
```diff
@@ -151,8 +151,7 @@ const AdminMenu: React.FC<MenuProps> = ({ children, ...menuProps }) => {
       ],
     },
     {
-      permissionIsAllowed:
-        !!enabledModules.podcast && (currentUserRole === 'app-owner' || currentUserRole === 'content-creator'),
+      permissionIsAllowed: !!enabledModules.podcast && currentUserRole === 'app-owner',
```
Comment (futianshen) [inline src/components/admin/AdminMenu.tsx:154]: should be recovered
`content-creator` has permission to use podcast program

File: src/components/common/ProductCreationModal.tsx [hunk 4]
```diff
@@ -99,7 +102,7 @@ const ProductCreationModal: React.FC<
           {formatMessage(commonMessages.ui.create)}
         </Button>
       )}
-      title={formatMessage(commonMessages.ui.create)}
+      title={customModalTitle ? customModalTitle : formatMessage(commonMessages.ui.create)}
```
Comment (futianshen) [inline src/components/common/ProductCreationModal.tsx:105]: NITS: customModalTitle || formatMessage(commonMessages.ui.create)

File: src/types/podcastAlbum.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { CategoryProps } from './general'
```
Comment (futianshen) [inline src/types/podcastAlbum.ts:1]: NITS: create another branch for refactor
`CategoryProps`  -> `Category`

File: src/types/podcast.ts [hunk 6]
```diff
@@ -20,6 +20,8 @@ export type PodcastProgramAdminProps = {
   audios: PodcastProgramAudio[]
 }
 
+export type PodcastProgramAdminProps = PodcastProgram
```
Comment (futianshen) [inline src/types/podcast.ts:23]: NITS: create another branch for refactor
`PodcastProgramAdminProps `  -> `PodcastProgram `

File: src/pages/PodcastAlbumCollectionAdminPage.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  const { currentMemberId } = useAuth()
+  const history = useHistory()
+  const [createPodcastAlbum] = useMutation<hasura.INSERT_PODCAST_ALBUM, hasura.INSERT_PODCAST_ALBUMVariables>(
+    INSERT_PODCAST_ALBUM,
+  )
+
+  // clean up
+  useEffect(() => {
+    return () => {}
+  }, [])
```
Comment (futianshen) [inline src/pages/PodcastAlbumCollectionAdminPage.tsx:31]: why need to clean up here ?

File: src/hooks/podcastAlbum.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    title: data?.podcast_album_by_pk?.title || '',
+    coverUrl: data?.podcast_album_by_pk?.cover_url || '',
+    author: {
+      id: data?.podcast_album_by_pk?.member.id || '',
+      name: data?.podcast_album_by_pk?.member.name || '',
+    },
+    description: data?.podcast_album_by_pk?.description || '',
+    isPublic: !!data?.podcast_album_by_pk?.is_public,
+    isDeleted: !!data?.podcast_album_by_pk?.is_deleted,
+    publishedAt: data?.podcast_album_by_pk?.published_at,
```
Comment (futianshen) [inline src/hooks/podcastAlbum.ts:54]: should be wrapped up with `new Date()`
```ts
 data?.podcast_album_by_pk?.published_at ? new Date( data?.podcast_album_by_pk?.published_at) | null
 ```

File: src/pages/PodcastAlbumCollectionAdminPage.tsx (hunk 7 above)
Comment (Ponchimeow) [reply src/pages/PodcastAlbumCollectionAdminPage.tsx:31]: try to fix
```
Warning: Can't perform a React state update on an unmounted component.
This is a no-op, but it indicates a memory leak in your application.
To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
```

File: src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    tab: string
+    condition: hasura.GET_PODCAST_ALBUM_PREVIEW_COLLECTIONVariables['condition']
+  }[] = [
+    {
+      key: 'published',
+      tab: formatMessage(commonMessages.status.published),
+      condition: {
+        published_at: { _is_null: false },
+        app_id: { _eq: appId },
+        is_deleted: { _eq: false },
```
Comment (futianshen) [inline src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx:27]: we've already set permission on `podcast_album`, not need to add these condition

File: src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+      },
+    },
+    {
+      key: 'draft',
+      tab: formatMessage(commonMessages.status.draft),
+      condition: {
+        published_at: { _is_null: true },
+        app_id: { _eq: appId },
+        is_deleted: { _eq: false },
```
Comment (futianshen) [inline src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx:36]: same as above

File: src/components/podcastAlbum/PodcastAlbumPodcastProgramConnectionModal.tsx [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+        const podcastProgramsIds = values.podcastProgramValues.map((value: string) => value.split('_')[0])
+
+        upsertPodcastAlbum({
+          variables: {
+            podcastPrograms: podcastProgramsIds.map((podcastProgramsId: string) => ({
+              podcast_album_id: podcastAlbumId,
+              podcast_program_id: podcastProgramsId,
+            })),
+            podcast_album_id: podcastAlbumId,
+            delete_podcast_programs_id: podcastPrograms
```
Comment (futianshen) [inline src/components/podcastAlbum/PodcastAlbumPodcastProgramConnectionModal.tsx:70]: avoid mixed naming convention

File: src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    key: string
+    tab: string
+    condition: hasura.GET_PODCAST_ALBUM_PREVIEW_COLLECTIONVariables['condition']
+  }[] = [
+    {
+      key: 'published',
+      tab: formatMessage(commonMessages.status.published),
+      condition: {
+        published_at: { _is_null: false },
+      },
```
Comment (futianshen) [inline src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx:26]: NITS: can directly use isPublished

File: src/pages/PodcastAlbumCollectionAdminPage.tsx (hunk 7 above)
Comment (kkshyu) [reply src/pages/PodcastAlbumCollectionAdminPage.tsx]: update the state of the unmounted component
```jsx
const [modal, setModal] = useState(true)

{
modal && <Modal renderTrigger={({ setVisible }) => {
     return <button onClick={() => {
        setModal(false)
        setTimeout(() => setVisible(true), 100)
    }}>Open</button>
}}>
}
```

File: src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx (hunk 9 above)
Comment (kkshyu) [reply src/components/podcastAlbum/PodcastAlbumCollectionTabs.tsx]: X-Hasura-User-Id: memberId
X-Hasura-App-Id: appId

```graphql
program_content(where: {
    program_content_section: {
        program: {
            app_id: { _eq: $appId }
            author_id: {_eq: $currentMemberId}
        }
    }
}) {
    id
    title
}
```

```sql
select program_content.id, program_content.title
from program_content
join program_content_section on ....
join program on ....
where program.app_id = $appId
```

```sql
with permitted_program_content as (
    select *
    from program_content
    where [permission]
) 
select id, title
from permitted_program_content
```

**Pull request merged**